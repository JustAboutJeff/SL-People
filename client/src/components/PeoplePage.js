import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Link } from 'react-router-dom'

const Page = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  height: 100%;
  width: 100%;
`
const TableWrap = styled.div`
  height: 37.5rem;
  overflow-y: auto;
  box-shadow: 0 0 6px 0 rgba(58, 58, 58, 0.24);
`

const Table = styled.table`
  width: 52rem;
  text-align: center;
`

const TableRow = styled.tr`
  padding: 1rem;
  transition: box-shadow 0.25s;

  &:nth-child(even) {
    background-color: #f0f8ff;
  }

  &:hover {
    box-shadow: inset 5px 0 0 0 rgba(0, 105, 224, 1),
      inset -5px 0 0 0 rgba(0, 105, 224, 1);
  }
`
const TableHeadRow = styled(TableRow)`
  font-size: 1.5rem;
  height: 3rem;
  color: white;
  font-weight: bold;
  text-align: center;
`

const TableHeader = styled.th`
  background-color: #4a90e2;
  box-shadow: 0 3px 0 0 rgba(0, 0, 0, 0.1);
  position: sticky;
  top: 0;
`

const StateButton = styled.button`
  align-items: center;
  background-color: #4a90e2;
  border-radius: 0.5rem;
  color: white;
  cursor: pointer;
  display: flex;
  font-size: 1.5rem;
  font-weight: bold;
  justify-content: center;
  min-width: 4rem;
  padding: 0.75rem;
  text-align: center;
`

const HomeButton = styled(Link)`
  align-items: center;
  background: #4a90e2;
  border-radius: 0.5rem;
  color: white;
  display: flex;
  font-size: 1.5rem;
  font-weight: bold;
  justify-content: center;
  min-width: 4rem;
  padding: 0.75rem;
  text-align: center;
`

const ButtonBar = styled.div`
  width: 52rem;
  margin-top: 15px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`

export class PeoplePage extends Component {
  constructor (props) {
    super(props)
    this.state = { showAnalysis: false }
    this.handleClick = this.handleClick.bind(this)
  }

  componentWillMount () {
    this.props.fetchPeople()
  }

  handleClick (e) {
    this.setState(prev => ({ showAnalysis: !prev.showAnalysis }))
  }

  renderLoading () {
    return <img src='/loading.gif' alt='Please wait your people are loading!' />
  }

  renderAnalysisRows (analysis) {
    return analysis.map(([character, count], idx) => (
      <TableRow key={idx}>
        <td>{character}</td>
        <td>{count}</td>
      </TableRow>
    ))
  }

  renderAnalysis (analysis) {
    return [
      <TableWrap key='1'>
        <Table>
          <thead>
            <TableHeadRow>
              <TableHeader>Character</TableHeader>
              <TableHeader>Count</TableHeader>
            </TableHeadRow>
          </thead>
          <tbody>{this.renderAnalysisRows(analysis)}</tbody>
        </Table>
      </TableWrap>,
      <ButtonBar key='2'>
        <HomeButton to={'/'}>Home</HomeButton>
        <StateButton onClick={this.handleClick}>People</StateButton>
      </ButtonBar>
    ]
  }

  renderPeopleRows (people) {
    return people.map((person, idx) => (
      <TableRow key={idx}>
        <td>{person.displayName}</td>
        <td>{person.emailAddress}</td>
        <td>{person.title}</td>
      </TableRow>
    ))
  }

  renderPeople (people) {
    return [
      <TableWrap key='1'>
        <Table>
          <thead>
            <TableHeadRow>
              <TableHeader>Name</TableHeader>
              <TableHeader>Email</TableHeader>
              <TableHeader>Title</TableHeader>
            </TableHeadRow>
          </thead>
          <tbody>{this.renderPeopleRows(people)}</tbody>
        </Table>
      </TableWrap>,
      <ButtonBar key='2'>
        <HomeButton to={'/'}>Home</HomeButton>
        <StateButton onClick={this.handleClick}>Analyze</StateButton>
      </ButtonBar>
    ]
  }

  render () {
    const { props: { analysis, people }, state: { showAnalysis } } = this

    let content = this.renderLoading()

    if (analysis.length && showAnalysis) {
      content = this.renderAnalysis(analysis)
    } else if (people.length) {
      content = this.renderPeople(people)
    }

    return <Page>{content}</Page>
  }
}

PeoplePage.propTypes = {
  analysis: PropTypes.array,
  people: PropTypes.array,
  fetchPeople: PropTypes.func
}

PeoplePage.defaultProps = {
  analysis: [],
  people: [],
  fetchPeople: Function.prototype
}

export default PeoplePage
