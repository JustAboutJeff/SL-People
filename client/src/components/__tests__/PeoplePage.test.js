import React, { isValidElement } from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { PeoplePage } from '../PeoplePage'

describe('PeoplePage', () => {
  let element,
    entry,
    renderer,
    props = {}
  beforeEach(() => {
    element = <PeoplePage {...props} />
    renderer = new ShallowRenderer()
  })

  it('returns a valid element', () => {
    expect(isValidElement(element)).toBe(true)
  })

  it('renders', () => {
    expect(renderer.render(element)).toMatchSnapshot()
  })

  describe('with people data', () => {
    props = {
      people: [
        {
          id: 101694867,
          displayName: 'Marisa Casper',
          emailAddress: 'isnaoj_nathz@ihooberbrunner.net',
          title: 'Direct Security Representative'
        },
        {
          id: 101694794,
          displayName: 'Griffin Hand',
          emailAddress: 'mamixe@lindgren.info',
          title: 'International Usability Agent'
        }
      ]
    }

    it('renders', () => {
      expect(renderer.render(element)).toMatchSnapshot()
    })
  })
})
