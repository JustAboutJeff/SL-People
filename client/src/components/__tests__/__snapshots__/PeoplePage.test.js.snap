// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeoplePage renders 1`] = `
<styled.div>
  <styled.div>
    <styled.table>
      <thead>
        <Styled(styled.tr)>
          <styled.th>
            Name
          </styled.th>
          <styled.th>
            Email
          </styled.th>
          <styled.th>
            Title
          </styled.th>
        </Styled(styled.tr)>
      </thead>
      <tbody>
        <styled.tr>
          <td>
            Marisa Casper
          </td>
          <td>
            isnaoj_nathz@ihooberbrunner.net
          </td>
          <td>
            Direct Security Representative
          </td>
        </styled.tr>
        <styled.tr>
          <td>
            Griffin Hand
          </td>
          <td>
            mamixe@lindgren.info
          </td>
          <td>
            International Usability Agent
          </td>
        </styled.tr>
      </tbody>
    </styled.table>
  </styled.div>
  <styled.div>
    <Styled(Link)
      to="/"
    >
      Home
    </Styled(Link)>
    <styled.button
      onClick={[Function]}
    >
      Analyze
    </styled.button>
  </styled.div>
</styled.div>
`;

exports[`PeoplePage with people data renders 1`] = `
<styled.div>
  <styled.div>
    <styled.table>
      <thead>
        <Styled(styled.tr)>
          <styled.th>
            Name
          </styled.th>
          <styled.th>
            Email
          </styled.th>
          <styled.th>
            Title
          </styled.th>
        </Styled(styled.tr)>
      </thead>
      <tbody>
        <styled.tr>
          <td>
            Marisa Casper
          </td>
          <td>
            isnaoj_nathz@ihooberbrunner.net
          </td>
          <td>
            Direct Security Representative
          </td>
        </styled.tr>
        <styled.tr>
          <td>
            Griffin Hand
          </td>
          <td>
            mamixe@lindgren.info
          </td>
          <td>
            International Usability Agent
          </td>
        </styled.tr>
      </tbody>
    </styled.table>
  </styled.div>
  <styled.div>
    <Styled(Link)
      to="/"
    >
      Home
    </Styled(Link)>
    <styled.button
      onClick={[Function]}
    >
      Analyze
    </styled.button>
  </styled.div>
</styled.div>
`;
